ralph.go

```diff
@@ -40,58 +45,395 @@ type MesosHost struct{}
+	if len(d.Create) == 0 && len(d.Update) == 0 && len(d.Delete) == 0 {
+		return true
+	}
+	return false
+}
+
+// SendDiffToRalph sends a given DiffEthernetComponent to Ralph. If dryRun is set to true,
+// then no changes will be sent to Ralph.
+func SendDiffToRalph(c *Client, d *DiffEthernetComponent, dryRun bool) error {
+	var msg = "EthernetComponent with MAC address %s %s successfully.\n"
+	var err error
+	for _, ec := range d.Create {
+		data, err := json.Marshal(ec)
+		if err != nil {
+			return err
+		}
+		if !dryRun {
+			err = c.SendToRalph("POST", APIEndpoints["EthernetComponent"], data)
+		}
+		if err != nil {
+			return err
+		}
+		fmt.Printf(msg, ec.MACAddress.String(), "created") // TODO(xor-xor): Use logger instead.
+	}
+	for _, ec := range d.Update {
+		data, err := json.Marshal(ec)
+		if err != nil {
+			return err
+		}
+		endpoint := fmt.Sprintf("%s/%d", APIEndpoints["EthernetComponent"], ec.ID)
+		if !dryRun {
+			err = c.SendToRalph("PUT", endpoint, data)
+		}
+		if err != nil {
+			return err
+		}
+		fmt.Printf(msg, ec.MACAddress.String(), "updated") // TODO(xor-xor): Use logger instead.
+	}
+	for _, ec := range d.Delete {
+		endpoint := fmt.Sprintf("%s/%d", APIEndpoints["EthernetComponent"], ec.ID)
+		if !dryRun {
+			err = c.SendToRalph("DELETE", endpoint, nil)
+		}
+		if err != nil {
+			return err
+		}
+		fmt.Printf(msg, ec.MACAddress.String(), "deleted") // TODO(xor-xor): Use logger instead.
+	}
+	return nil
+}
+
+// CompareEthernetComponents compares two sets of EthernetComponents (old and new) and
+// creates DiffEthernetComponent holding detected changes.
+func CompareEthernetComponents(old, new []*EthernetComponent) (*DiffEthernetComponent, error) {
+	var all, create, update, delete []*EthernetComponent
+	all = append(all, old...)
+	all = append(all, new...)
+
+	for _, eth := range all {
+		switch mac := eth.MACAddress; {
+		case contains(new, mac) && !contains(old, mac):
+			create = append(create, eth)
+		case contains(old, mac) && !contains(new, mac):
+			delete = append(delete, eth)
+		default:
+			// We need to find matching eth in old and assign its ID to the new one
+			// to make an update. Also, we need to exclude eth equal to ethOld (there's
+			// no need to update in such case).
+			if eth.ID == 0 {
+				var ethOld *EthernetComponent
+				for _, ethOld = range old {
+					if ethOld.MACAddress.String() == eth.MACAddress.String() {
+						eth.ID = ethOld.ID
+						break
+					}
+				}
+				if eth.ID == 0 {
+					// This really shouldn't happen.
+					return nil, fmt.Errorf("error while preparing EthernetComponent with MAC %s for update", eth.MACAddress)
+				}
+				if !eth.IsEqualTo(ethOld) {
+					update = append(update, eth)
+				}
+			}
+		}
+	}
+	return &DiffEthernetComponent{
+		Create: create,
+		Delete: delete,
+		Update: update,
+	}, nil
+}
+
+// IsEqualTo compares two EthernetComponents for equality.
+func (ec1 *EthernetComponent) IsEqualTo(ec2 *EthernetComponent) bool {
```

Comment:
Because reflection-based functions are not for casual use, i.e. they should be used only when there's no easy way to implement desired functionality, which obviously is not the case here (implementation of `IsEqualTo` is kind of trivial).

There's a whole section devoted to this topic in "The Go Programming Language" by Donovan & Kernighan, if you're interested (see: 12.9. "A Word of Caution").